<a href="https://colab.research.google.com/github/dayeong089/web_scrapping/blob/main/Week3_Pandas_Dataframes_Homework.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Software Coaching for Python
# Week 3: Pandas Dataframes - Homework

Instructor: Kang-Pyo Lee

In [ ]:
from google.colab import drive

drive.mount('/content/drive')
path = "/content/drive/My Drive/PythonCoaching"
my_folder = "dayoung21"     # *** REPLACE WITH YOUR FOLDER NAME ***
outcome_folder = f"{path}/{my_folder}/outcome"
classdata_folder = f"{path}/classdata"

Mounted at /content/drive


## Element Selection from a Dataframe

In [ ]:
import pandas as pd
pd.set_option('display.max_colwidth', 150)

df = pd.read_csv(f"{classdata_folder}/timeline_cnnbrk.csv", sep="\t")
df

status_id                      created_at  \
0     1312243714585235457  Sat Oct 03 04:10:56 +0000 2020   
1     1312217447752237056  Sat Oct 03 02:26:34 +0000 2020   
2     1312189828247691265  Sat Oct 03 00:36:49 +0000 2020   
3     1312188613015203842  Sat Oct 03 00:31:59 +0000 2020   
4     1312161028923244545  Fri Oct 02 22:42:22 +0000 2020   
...                   ...                             ...   
3195  1231426815996170243  Sun Feb 23 03:53:26 +0000 2020   
3196  1231402873617108992  Sun Feb 23 02:18:18 +0000 2020   
3197  1231385843723636739  Sun Feb 23 01:10:37 +0000 2020   
3198  1231338606863601668  Sat Feb 22 22:02:55 +0000 2020   
3199  1231307966487449600  Sat Feb 22 20:01:10 +0000 2020   

                                                                                                                                              text  \
0     President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA   
1     Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
2     Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp   
3     After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC   
4      President Trump has landed on Marine One at Walter Reed medical center, where he will be hospitalized for the next… https://t.co/Btb0FBkJLW   
...                                                                                                                                            ...   
3195                   Science Channel star Mad Mike Hughes died Saturday during an attempt to launch his homemade rocket… https://t.co/ezI1BUHstT   
3196  A person was killed in New Orleans when they were hit by a float during the Krewe of Endymion parade, the city says… https://t.co/wC9pSiSqTv   
3197  Sen. Bernie Sanders will win a decisive victory in Nevada's Democratic caucuses, CNN projects, solidifying his posi… https://t.co/Wk2nFjEiOw   
3198   Three people were killed and 18 injured after a bus rolled over on Interstate 15 near Fallbrook, California, north… https://t.co/UHqTVpV7Vm   
3199                                                     Nevada's Democratic caucuses are underway https://t.co/7O0yKSUzzI https://t.co/321iXf34lS   

      retweet_count  favorite_count  is_retweet retweet_created_at  
0              1050            3946           0                NaN  
1              2287            7232           0                NaN  
2               501            1559           0                NaN  
3              2831            5992           0                NaN  
4               249             841           0                NaN  
...             ...             ...         ...                ...  
3195            345             961           0                NaN  
3196            194             345           0                NaN  
3197            321            2058           0                NaN  
3198             97             305           0                NaN  
3199             29             196           0                NaN  

[3200 rows x 7 columns]

1\. Get the list of column labels of `df`.

In [ ]:
# Your answer here
df.columns

Index(['status_id', 'created_at', 'text', 'retweet_count', 'favorite_count',
       'is_retweet', 'retweet_created_at'],
      dtype='object')

2\. Get the list of row index positions of `df`.

In [ ]:
# Your answer here
list(df.index)

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


3\. Get the shape of `df`.

In [ ]:
# Your answer here
df.shape

(3200, 7)

4\. Get the number of rows in `df`.

In [ ]:
# Your answer here
len(df)

3200

5\. Select all values from the `created_at` column.

In [ ]:
# Your answer here
df.created_at

0       Sat Oct 03 04:10:56 +0000 2020
1       Sat Oct 03 02:26:34 +0000 2020
2       Sat Oct 03 00:36:49 +0000 2020
3       Sat Oct 03 00:31:59 +0000 2020
4       Fri Oct 02 22:42:22 +0000 2020
                     ...              
3195    Sun Feb 23 03:53:26 +0000 2020
3196    Sun Feb 23 02:18:18 +0000 2020
3197    Sun Feb 23 01:10:37 +0000 2020
3198    Sat Feb 22 22:02:55 +0000 2020
3199    Sat Feb 22 20:01:10 +0000 2020
Name: created_at, Length: 3200, dtype: object

6\. Select the first 3 rows from `df`.

In [ ]:
# Your answer here
df[:3]

status_id                      created_at  \
0  1312243714585235457  Sat Oct 03 04:10:56 +0000 2020   
1  1312217447752237056  Sat Oct 03 02:26:34 +0000 2020   
2  1312189828247691265  Sat Oct 03 00:36:49 +0000 2020   

                                                                                                                                           text  \
0  President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA   
1  Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
2  Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp   

   retweet_count  favorite_count  is_retweet retweet_created_at  
0           1050            3946           0                NaN  
1           2287            7232           0                NaN  
2            501            1559           0                NaN

7\. Select the last 3 rows from `df`.

In [ ]:
# Your answer here
df[-3:]

status_id                      created_at  \
3197  1231385843723636739  Sun Feb 23 01:10:37 +0000 2020   
3198  1231338606863601668  Sat Feb 22 22:02:55 +0000 2020   
3199  1231307966487449600  Sat Feb 22 20:01:10 +0000 2020   

                                                                                                                                              text  \
3197  Sen. Bernie Sanders will win a decisive victory in Nevada's Democratic caucuses, CNN projects, solidifying his posi… https://t.co/Wk2nFjEiOw   
3198   Three people were killed and 18 injured after a bus rolled over on Interstate 15 near Fallbrook, California, north… https://t.co/UHqTVpV7Vm   
3199                                                     Nevada's Democratic caucuses are underway https://t.co/7O0yKSUzzI https://t.co/321iXf34lS   

      retweet_count  favorite_count  is_retweet retweet_created_at  
3197            321            2058           0                NaN  
3198             97             305           0                NaN  
3199             29             196           0                NaN

8\. Select the element at row index position 999 and in the `text` column.

In [ ]:
# Your answer here
df["text"][999]

"US agents, local law enforcement and a locksmith's van entered the Chinese consulate in Houston days after it was o… https://t.co/25xHJptVgE"

9\. Select the element in the fifth row and the third column.

In [ ]:
# Your answer here
df.iloc[4, 2]

'President Trump has landed on Marine One at Walter Reed medical center, where he will be hospitalized for the next… https://t.co/Btb0FBkJLW'

10\. Select all values from the last column, without specifying the column label.

In [ ]:
# Your answer here
df.iloc[:, -1]

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
3195    NaN
3196    NaN
3197    NaN
3198    NaN
3199    NaN
Name: retweet_created_at, Length: 3200, dtype: object

11\. Select all rows in the last two columns, without specifying the column labels.

In [ ]:
# Your answer here
df.iloc[:, -2:]

is_retweet retweet_created_at
0              0                NaN
1              0                NaN
2              0                NaN
3              0                NaN
4              0                NaN
...          ...                ...
3195           0                NaN
3196           0                NaN
3197           0                NaN
3198           0                NaN
3199           0                NaN

[3200 rows x 2 columns]

12\. Select the first five rows in the last two columns, without specifying the column labels.

In [ ]:
# Your answer here
df.iloc[:5, -2:]

is_retweet retweet_created_at
0           0                NaN
1           0                NaN
2           0                NaN
3           0                NaN
4           0                NaN

13\. Select all rows from `df` with the `favorite_count` column value being strictly larger than 5000.

In [ ]:
# Your answer here
df[df.favorite_count > 5000]

status_id                      created_at  \
1     1312217447752237056  Sat Oct 03 02:26:34 +0000 2020   
3     1312188613015203842  Sat Oct 03 00:31:59 +0000 2020   
5     1312155447839780864  Fri Oct 02 22:20:12 +0000 2020   
17    1312067092980604928  Fri Oct 02 16:29:06 +0000 2020   
23    1311898474875703296  Fri Oct 02 05:19:04 +0000 2020   
...                   ...                             ...   
2933  1238646170307760128  Sat Mar 14 02:00:34 +0000 2020   
2951  1238286757264797697  Fri Mar 13 02:12:23 +0000 2020   
2953  1238259897743138816  Fri Mar 13 00:25:40 +0000 2020   
3009  1237395474371481600  Tue Mar 10 15:10:45 +0000 2020   
3036  1236432293906075651  Sat Mar 07 23:23:25 +0000 2020   

                                                                                                                                              text  \
1     Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
3     After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC   
5     The Biden campaign is pulling all negative ads after President Trump was diagnosed with Covid-19, a campaign aide s… https://t.co/iaxbszv9Zn   
17    Presidential nominee Joe Biden and his wife, Jill Biden, tested negative for coronavirus this morning, their doctor… https://t.co/ryLBhUByku   
23                          President Trump tweets that he and first lady Melania Trump have tested positive for Covid-19… https://t.co/didusJ091s   
...                                                                                                                                            ...   
2933  Google will not be publishing a national-scale website for coronavirus testing anytime soon, contrary to claims mad… https://t.co/xSz4KsBDXw   
2951  Canadian Prime Minister Justin Trudeau's wife Sophie has tested positive for coronavirus. The Prime Minister does n… https://t.co/lo6yZLeTsk   
2953  Disney is closing Walt Disney World, its flagship theme park resort in Florida, because of the coronavirus outbreak… https://t.co/2uWdlHQRaq   
3009                               Dick's Sporting Goods will stop selling guns at 440 more stores https://t.co/sYV3zPn7k3 https://t.co/OZpLeAAijM   
3036  A person who attended CPAC last week tested positive for coronavirus. Many officials, including Trump and Pence, we… https://t.co/31hH9vlDL7   

      retweet_count  favorite_count  is_retweet retweet_created_at  
1              2287            7232           0                NaN  
3              2831            5992           0                NaN  
5               948            8089           0                NaN  
17              894            7653           0                NaN  
23             2466            9103           0                NaN  
...             ...             ...         ...                ...  
2933           4348            9560           0                NaN  
2951           3413            6179           0                NaN  
2953           3172            5426           0                NaN  
3009            774            7854           0                NaN  
3036           4434           12796           0                NaN  

[195 rows x 7 columns]

14\. Select all rows from `df` with the `favorite_count` column value being between 5000 (inclusive) and 10000 (exclusive).

In [ ]:
# Your answer here
df[(df.favorite_count >= 5000) & (df.favorite_count < 10000)]

status_id                      created_at  \
1     1312217447752237056  Sat Oct 03 02:26:34 +0000 2020   
3     1312188613015203842  Sat Oct 03 00:31:59 +0000 2020   
5     1312155447839780864  Fri Oct 02 22:20:12 +0000 2020   
17    1312067092980604928  Fri Oct 02 16:29:06 +0000 2020   
23    1311898474875703296  Fri Oct 02 05:19:04 +0000 2020   
...                   ...                             ...   
2822  1240482077713334272  Thu Mar 19 03:35:49 +0000 2020   
2933  1238646170307760128  Sat Mar 14 02:00:34 +0000 2020   
2951  1238286757264797697  Fri Mar 13 02:12:23 +0000 2020   
2953  1238259897743138816  Fri Mar 13 00:25:40 +0000 2020   
3009  1237395474371481600  Tue Mar 10 15:10:45 +0000 2020   

                                                                                                                                              text  \
1     Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
3     After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC   
5     The Biden campaign is pulling all negative ads after President Trump was diagnosed with Covid-19, a campaign aide s… https://t.co/iaxbszv9Zn   
17    Presidential nominee Joe Biden and his wife, Jill Biden, tested negative for coronavirus this morning, their doctor… https://t.co/ryLBhUByku   
23                          President Trump tweets that he and first lady Melania Trump have tested positive for Covid-19… https://t.co/didusJ091s   
...                                                                                                                                            ...   
2822  For the first time since the novel coronavirus outbreak began late last year, China has reported no new local infec… https://t.co/KAHAmEbhi9   
2933  Google will not be publishing a national-scale website for coronavirus testing anytime soon, contrary to claims mad… https://t.co/xSz4KsBDXw   
2951  Canadian Prime Minister Justin Trudeau's wife Sophie has tested positive for coronavirus. The Prime Minister does n… https://t.co/lo6yZLeTsk   
2953  Disney is closing Walt Disney World, its flagship theme park resort in Florida, because of the coronavirus outbreak… https://t.co/2uWdlHQRaq   
3009                               Dick's Sporting Goods will stop selling guns at 440 more stores https://t.co/sYV3zPn7k3 https://t.co/OZpLeAAijM   

      retweet_count  favorite_count  is_retweet retweet_created_at  
1              2287            7232           0                NaN  
3              2831            5992           0                NaN  
5               948            8089           0                NaN  
17              894            7653           0                NaN  
23             2466            9103           0                NaN  
...             ...             ...         ...                ...  
2822           1438            7417           0                NaN  
2933           4348            9560           0                NaN  
2951           3413            6179           0                NaN  
2953           3172            5426           0                NaN  
3009            774            7854           0                NaN  

[141 rows x 7 columns]

15\. Create a random sample of 100 rows from `df` without duplicates.

In [ ]:
# Your answer here
df.sample(n=100, replace=False)

status_id                      created_at  \
2434  1252956638274891776  Wed Apr 22 13:45:16 +0000 2020   
2815  1240647993952546817  Thu Mar 19 14:35:06 +0000 2020   
1885  1266906397385592839  Sun May 31 01:36:37 +0000 2020   
1036  1285511149761581056  Tue Jul 21 09:45:16 +0000 2020   
1646  1271060469118353408  Thu Jun 11 12:43:25 +0000 2020   
...                   ...                             ...   
2950  1238302602389053441  Fri Mar 13 03:15:21 +0000 2020   
2456  1252414166410780675  Tue Apr 21 01:49:40 +0000 2020   
734   1293987784450072576  Thu Aug 13 19:08:23 +0000 2020   
1122  1283146972652871681  Tue Jul 14 21:10:52 +0000 2020   
847   1290752931793375232  Tue Aug 04 20:54:14 +0000 2020   

                                                                                                                                              text  \
2434    RT @CNNPolitics: President Trump on Twitter tells the US Navy to shoot down and destroy Iranian boats that harass US ships https://t.co/p…   
2815     Two TSA baggage screening officers at New York's JFK airport test positive for coronavirus. Follow live updates:… https://t.co/N29gpZZncm   
1885                      Texas and Colorado have activated the National Guard respond to protests https://t.co/vTqLcHE80k https://t.co/tIckwY33Zz   
1036  The United States is restricting 11 Chinese companies from buying American technology and other goods because it sa… https://t.co/6LqU8OlBw6   
1646  A medicine that may treat and prevent Covid-19 is now being tested in patients in multiple sites around the US, Reg… https://t.co/tB4cYMKBaP   
...                                                                                                                                            ...   
2950  The president of the US Soccer Federation has stepped down and apologized for the wording in the organization's leg… https://t.co/Jqh0wXQuJ2   
2456  US is monitoring intelligence that North Korean leader Kim Jong Un is in grave danger after surgery, according to a… https://t.co/WtYggnci5t   
734   A federal judge in Pennsylvania told the Trump campaign and the Republican Party that they must produce evidence th… https://t.co/RAqT48hsUk   
1122  The first phase of trials for a Covid-19 vaccine showed that it worked to trigger an immune response with mild side… https://t.co/CZ1ctpoH9J   
847   A bipartisan group of six governors has signed a deal that would help create the country's first coordinated testin… https://t.co/fB78fivQTM   

      retweet_count  favorite_count  is_retweet  \
2434            321               0           1   
2815            369             494           0   
1885            478             813           0   
1036            411            1197           0   
1646            495            1497           0   
...             ...             ...         ...   
2950            231            1170           0   
2456           4181            7491           0   
734            1356            5850           0   
1122            545            1778           0   
847             609            3018           0   

                  retweet_created_at  
2434  Wed Apr 22 13:07:11 +0000 2020  
2815                             NaN  
1885                             NaN  
1036                             NaN  
1646                             NaN  
...                              ...  
2950                             NaN  
2456                             NaN  
734                              NaN  
1122                             NaN  
847                              NaN  

[100 rows x 7 columns]

## Dataframe Iteration & File Writing

Let's continue to use `df`.

1\. Select the first ten values from the `text` column of `df` and print the first 30 characters of each value concatenated by the string '...'. Rows in the printed output are separated by a new line. For example, the first three rows in the printed output look as follows.

President Donald Trump's campa...<br>
Former White House adviser Kel...<br>
Sen. Thom Tillis says he has t...<br>

In [ ]:
# Your answer here
for text in df[:10].text:
    print(text[:30] + "...")

President Donald Trump's campa...
Former White House adviser Kel...
Sen. Thom Tillis says he has t...
After Joe Biden's campaign sai...
President Trump has landed on ...
The Biden campaign is pulling ...
The White House says President...
President Trump is headed to W...
President Trump has had a feve...
Disgraced movie mogul Harvey W...


2\. Select the first ten rows from `df` and print the values of each row in the following format: 'created at: CREATED_AT, text: TEXT, retweet_count: RETWEET_COUNT', where CREATED_AT, TEXT, and RETWEET_COUNT are the values from the three columns `created_at`, `text`, and `retweet_count`, respectively. Rows are separated by two new lines. For example, the first three rows in the printed output look as follows.

created_at: Sat Oct 03 04:10:56 +0000 2020, text: President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA, retweet_count: 1050<br>
<br>
created_at: Sat Oct 03 02:26:34 +0000 2020, text: Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj, retweet_count: 2287<br>
<br>
created_at: Sat Oct 03 00:36:49 +0000 2020, text: Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp, retweet_count: 501<br>

In [ ]:
# Your answer here
for idx, row in df[:10].iterrows():
  created = row.created_at
  text = row.text
  retweet = row.retweet_count
  print(f"created_at:{created}, text:{text}, retweet_count:{retweet}\n\n", end="")

created_at:Sat Oct 03 04:10:56 +0000 2020, text:President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA, retweet_count:1050

created_at:Sat Oct 03 02:26:34 +0000 2020, text:Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj, retweet_count:2287

created_at:Sat Oct 03 00:36:49 +0000 2020, text:Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp, retweet_count:501

created_at:Sat Oct 03 00:31:59 +0000 2020, text:After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC, retweet_count:2831

created_at:Fri Oct 02 22:42:22 +0000 2020, text:President Trump has landed on Marine One at Walter Reed medical center, where he will be hospitalized for the nex

3\. Select all rows from `df` with the `is_retweet` column value being 0 and save the resulting dataframe in a CSV file named hw3_output.csv in the `outcome` folder such that the column separator is a tab and it contains the index of the dataframe.

In [ ]:
# Your answer here
df2 = df[df.is_retweet == 0]
df2

status_id                      created_at  \
0     1312243714585235457  Sat Oct 03 04:10:56 +0000 2020   
1     1312217447752237056  Sat Oct 03 02:26:34 +0000 2020   
2     1312189828247691265  Sat Oct 03 00:36:49 +0000 2020   
3     1312188613015203842  Sat Oct 03 00:31:59 +0000 2020   
4     1312161028923244545  Fri Oct 02 22:42:22 +0000 2020   
...                   ...                             ...   
3195  1231426815996170243  Sun Feb 23 03:53:26 +0000 2020   
3196  1231402873617108992  Sun Feb 23 02:18:18 +0000 2020   
3197  1231385843723636739  Sun Feb 23 01:10:37 +0000 2020   
3198  1231338606863601668  Sat Feb 22 22:02:55 +0000 2020   
3199  1231307966487449600  Sat Feb 22 20:01:10 +0000 2020   

                                                                                                                                              text  \
0     President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA   
1     Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
2     Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp   
3     After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC   
4      President Trump has landed on Marine One at Walter Reed medical center, where he will be hospitalized for the next… https://t.co/Btb0FBkJLW   
...                                                                                                                                            ...   
3195                   Science Channel star Mad Mike Hughes died Saturday during an attempt to launch his homemade rocket… https://t.co/ezI1BUHstT   
3196  A person was killed in New Orleans when they were hit by a float during the Krewe of Endymion parade, the city says… https://t.co/wC9pSiSqTv   
3197  Sen. Bernie Sanders will win a decisive victory in Nevada's Democratic caucuses, CNN projects, solidifying his posi… https://t.co/Wk2nFjEiOw   
3198   Three people were killed and 18 injured after a bus rolled over on Interstate 15 near Fallbrook, California, north… https://t.co/UHqTVpV7Vm   
3199                                                     Nevada's Democratic caucuses are underway https://t.co/7O0yKSUzzI https://t.co/321iXf34lS   

      retweet_count  favorite_count  is_retweet retweet_created_at  
0              1050            3946           0                NaN  
1              2287            7232           0                NaN  
2               501            1559           0                NaN  
3              2831            5992           0                NaN  
4               249             841           0                NaN  
...             ...             ...         ...                ...  
3195            345             961           0                NaN  
3196            194             345           0                NaN  
3197            321            2058           0                NaN  
3198             97             305           0                NaN  
3199             29             196           0                NaN  

[2852 rows x 7 columns]

In [ ]:
df2.to_csv(f"{outcome_folder}/hw3_output.csv", sep="\t", index=True)

## Adding Columns to a Dataframe

Let's continue to use `df`.

1\. Add a new column `text_shortened` to `df` such that each value in the column is the first 30 characters of the `text` column value concatenated by the string '...'.  

In [ ]:
# Your answer here
df["text_shortened"] = df.text.apply(lambda x: x[:30] + "...")

In [ ]:
# Check your answer here.
df[["text", "text_shortened"]]

text  \
0     President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA   
1     Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
2     Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp   
3     After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC   
4      President Trump has landed on Marine One at Walter Reed medical center, where he will be hospitalized for the next… https://t.co/Btb0FBkJLW   
...                                                                                                                                            ...   
3195                   Science Channel star Mad Mike Hughes died Saturday during an attempt to launch his homemade rocket… https://t.co/ezI1BUHstT   
3196  A person was killed in New Orleans when they were hit by a float during the Krewe of Endymion parade, the city says… https://t.co/wC9pSiSqTv   
3197  Sen. Bernie Sanders will win a decisive victory in Nevada's Democratic caucuses, CNN projects, solidifying his posi… https://t.co/Wk2nFjEiOw   
3198   Three people were killed and 18 injured after a bus rolled over on Interstate 15 near Fallbrook, California, north… https://t.co/UHqTVpV7Vm   
3199                                                     Nevada's Democratic caucuses are underway https://t.co/7O0yKSUzzI https://t.co/321iXf34lS   

                         text_shortened  
0     President Donald Trump's campa...  
1     Former White House adviser Kel...  
2     Sen. Thom Tillis says he has t...  
3     After Joe Biden's campaign sai...  
4     President Trump has landed on ...  
...                                 ...  
3195  Science Channel star Mad Mike ...  
3196  A person was killed in New Orl...  
3197  Sen. Bernie Sanders will win a...  
3198  Three people were killed and 1...  
3199  Nevada's Democratic caucuses a...  

[3200 rows x 2 columns]

2\. Add a new column `first_last_token` to `df` such that each column value is a tuple with two elements: the first element being the first token of the `text` column value and the second one being the last token (use the <b>split</b> method).

In [ ]:
# Your answer here
df["first_last_token"] = df.text.apply(lambda x: (x.split()[0], x.split()[-1]))

In [ ]:
# Check your answer here.
df[["text", "first_last_token"]]

text  \
0     President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA   
1     Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
2     Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp   
3     After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC   
4      President Trump has landed on Marine One at Walter Reed medical center, where he will be hospitalized for the next… https://t.co/Btb0FBkJLW   
...                                                                                                                                            ...   
3195                   Science Channel star Mad Mike Hughes died Saturday during an attempt to launch his homemade rocket… https://t.co/ezI1BUHstT   
3196  A person was killed in New Orleans when they were hit by a float during the Krewe of Endymion parade, the city says… https://t.co/wC9pSiSqTv   
3197  Sen. Bernie Sanders will win a decisive victory in Nevada's Democratic caucuses, CNN projects, solidifying his posi… https://t.co/Wk2nFjEiOw   
3198   Three people were killed and 18 injured after a bus rolled over on Interstate 15 near Fallbrook, California, north… https://t.co/UHqTVpV7Vm   
3199                                                     Nevada's Democratic caucuses are underway https://t.co/7O0yKSUzzI https://t.co/321iXf34lS   

                          first_last_token  
0     (President, https://t.co/piwok6WXpA)  
1        (Former, https://t.co/DzHyaaO7Cj)  
2          (Sen., https://t.co/J3cnZUTEjp)  
3         (After, https://t.co/hArgCeKRzC)  
4     (President, https://t.co/Btb0FBkJLW)  
...                                    ...  
3195    (Science, https://t.co/ezI1BUHstT)  
3196          (A, https://t.co/wC9pSiSqTv)  
3197       (Sen., https://t.co/Wk2nFjEiOw)  
3198      (Three, https://t.co/UHqTVpV7Vm)  
3199   (Nevada's, https://t.co/321iXf34lS)  

[3200 rows x 2 columns]

In [ ]:
type(df.first_last_token[0])

tuple

3\. Add a new column `metrics` to `df` such that each column value is a tuple with three values from the three columns `retweet_count`, `favorite_count`, and `is_retweet`, respectively, in order.

In [ ]:
# Your answer here
df["metrics"] = df.apply(lambda x: (x.retweet_count, x.favorite_count, x.is_retweet), axis=1)

In [ ]:
# Check your answer here.
df[["retweet_count", "favorite_count", "is_retweet", "metrics"]]

retweet_count  favorite_count  is_retweet          metrics
0              1050            3946           0  (1050, 3946, 0)
1              2287            7232           0  (2287, 7232, 0)
2               501            1559           0   (501, 1559, 0)
3              2831            5992           0  (2831, 5992, 0)
4               249             841           0    (249, 841, 0)
...             ...             ...         ...              ...
3195            345             961           0    (345, 961, 0)
3196            194             345           0    (194, 345, 0)
3197            321            2058           0   (321, 2058, 0)
3198             97             305           0     (97, 305, 0)
3199             29             196           0     (29, 196, 0)

[3200 rows x 4 columns]

4\. Add a new column `num_mentions` to `df` such that each column value is the number of '@' symbols in the `text` column value.

In [ ]:
# Your answer here
df["num_mentions"] = df.text.str.count("@")

In [ ]:
# Check your answer here.
df[["text", "num_mentions"]]

text  \
0     President Donald Trump's campaign manager Bill Stepien has tested positive for coronavirus, according to a senior o… https://t.co/piwok6WXpA   
1     Former White House adviser Kellyanne Conway says she has tested positive for Covid-19. She attended a Rose Garden e… https://t.co/DzHyaaO7Cj   
2     Sen. Thom Tillis says he has tested positive for Covid-19. He's the second member of the Senate Judiciary Committee… https://t.co/J3cnZUTEjp   
3     After Joe Biden's campaign said it would pull negative ads following President Trump's Covid-19 diagnosis, the Pres… https://t.co/hArgCeKRzC   
4      President Trump has landed on Marine One at Walter Reed medical center, where he will be hospitalized for the next… https://t.co/Btb0FBkJLW   
...                                                                                                                                            ...   
3195                   Science Channel star Mad Mike Hughes died Saturday during an attempt to launch his homemade rocket… https://t.co/ezI1BUHstT   
3196  A person was killed in New Orleans when they were hit by a float during the Krewe of Endymion parade, the city says… https://t.co/wC9pSiSqTv   
3197  Sen. Bernie Sanders will win a decisive victory in Nevada's Democratic caucuses, CNN projects, solidifying his posi… https://t.co/Wk2nFjEiOw   
3198   Three people were killed and 18 injured after a bus rolled over on Interstate 15 near Fallbrook, California, north… https://t.co/UHqTVpV7Vm   
3199                                                     Nevada's Democratic caucuses are underway https://t.co/7O0yKSUzzI https://t.co/321iXf34lS   

      num_mentions  
0                0  
1                0  
2                0  
3                0  
4                0  
...            ...  
3195             0  
3196             0  
3197             0  
3198             0  
3199             0  

[3200 rows x 2 columns]

In [ ]:
df.num_mentions.value_counts()

0    2852
1     346
2       2
Name: num_mentions, dtype: int64

In [ ]:
df.columns

Index(['status_id', 'created_at', 'text', 'retweet_count', 'favorite_count',
       'is_retweet', 'retweet_created_at', 'text_shortened',
       'first_last_token', 'metrics', 'num_mentions'],
      dtype='object')